frontend/src/metabase/query_builder/components/NativeQueryEditor/AceEditor.tsx

```diff
@@ -0,0 +1,568 @@
+          const cacheHit =
+            Date.now() - timestamp < AUTOCOMPLETE_CACHE_DURATION &&
+            lastAutoComplete.prefix === prefix;
+
+          if (!cacheHit) {
+            // Get models and fields from tables
+            // HACK: call this.props.autocompleteResultsFn rather than caching the prop since it might change
+            const apiResults = await this.props.autocompleteResultsFn(prefix);
+            lastAutoComplete = {
+              timestamp: Date.now(),
+              prefix,
+              results: apiResults,
+            };
+
+            // Get referenced questions
+            const referencedQuestionIds =
+              this.props.query.referencedQuestionIds();
+            // The results of the API call are cached by ID
+            const referencedCards = await Promise.all(
+              referencedQuestionIds.map(id => this.props.fetchQuestion(id)),
+            );
+
+            // Get columns from referenced questions that match the prefix
+            const lowerCasePrefix = prefix.toLowerCase();
+            const isMatchForPrefix = (name: string) =>
+              name.toLowerCase().includes(lowerCasePrefix);
+            const questionColumns: AutocompleteItem[] = referencedCards
+              .filter(Boolean)
+              .flatMap(card =>
+                card.result_metadata
+                  .filter(columnMetadata =>
+                    isMatchForPrefix(columnMetadata.name),
+                  )
+                  .map(
+                    columnMetadata =>
+                      [
+                        columnMetadata.name,
+                        `${card.name} :${columnMetadata.base_type}`,
+                      ] as AutocompleteItem,
+                  ),
+              );
+
+            // Concat the results from tables, fields, and referenced questions.
+            // The ace editor will deduplicate results based on name, keeping results
+            // that come first. In case of a name conflict, prioritise referenced
+            // questions' columns over tables and fields.
+            results = questionColumns.concat(apiResults);
+          }
+
+          // transform results into what ACE expects
+          callback(null, prepareResultsForAce(results));
+        } catch (error) {
+          console.error("error getting autocompletion data", error);
+          callback(null, []);
+        }
+      },
+    });
+
+    // the completers when the editor mounts are the standard ones
+    const standardCompleters = [...this._editor.completers];
+
+    this.nextCompleters = pos => {
+      if (this.getSnippetNameAtCursor(pos)) {
+        return [{ getCompletions: this.getSnippetCompletions }];
+      } else if (this.getCardTagNameAtCursor(pos)) {
+        return [{ getCompletions: this.getCardTagCompletions }];
+      } else {
+        return [...standardCompleters];
+      }
+    };
+
+    this.HACK_initializeCompletionsPopup();
+  }
+
+  HACK_initializeCompletionsPopup = () => {
```

Comment:
ACE does not render the autocomplete popover in it's wrapping container. This means we have had to use global css for styling the autocomplete popover and it's contents.

This hacky bit of code adds a custom class to the autocomplete popover when it renders, so we don't need to rely on global css anymore.